src/python/deepgnn/graph_engine/snark/synchronized.py

```diff
@@ -0,0 +1,49 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT License.
+"""Synchronized server - client setup."""
+from deepgnn.graph_engine.snark.distributed import Server, Client as DistributedClient
+import ray
+
+
+@ray.remote
+class ServerWrapper(object):
+    def __init__(self):
+        self.server = Server("localhost:9999", "/tmp/cora", 0, 1)
+
+    def reset(self):
+        self.server.reset()
+
+    def get_hostname(self):
+        return self.server._hostname
+
+
+@ray.remote
+def start_servers(num: int):
+    return [ServerWrapper.remote() for i in range(num)]
+
+
+@ray.remote
+def start_clients(servers, n_clients):
+    return [
+        DistributedClient([ray.get(server.get_hostname.remote()) for server in servers])
+        for _ in range(n_clients)
+    ]
+
+
+@ray.remote
+def train(servers, clients):
+
+    for cl in clients:
+        cl.reset()
+    for s in servers:
+        ray.get(s.reset.remote())
+    return
+
+
+servers = start_servers.bind(2)
+clients = start_clients.bind(servers, 3)
+output = train.bind(servers, clients)
+
+from ray import workflow
+
+print(workflow.run(output))
```

Comment:
Here is the workflow example you are looking for - start 2 servers, connect with 3 clients then train.

Reply:
1. I don't think it is in an example folder..
2.  There is no training..
3. > We definitely don't want the graph engine package to depend on ray.

Reply:
Can you review it to make sure the design is ok before I add training and a whole example?

Reply:
Could you verify that the proposed design will work on tests from:
https://github.com/microsoft/DeepGNN/blob/main/src/python/deepgnn/graph_engine/backends/snark/test_snark_client.py
first? Then try to make sure, that *graph engine package will not depend on ray* for every use case and we have an example, how to use it in a distributed setting.

Reply:
It looks like the design will work on that old unit test, the new unit tests are in test_synchronized. As far as depending on Ray, the synchronized code depends on Ray since it uses Ray workflow but nothing else.

Reply:
Sounds good then. Try to autodetect host names for different settings listed in that test and graduate this PR from wip.

Reply:
I don't see this:
> Try to autodetect host names for different settings listed in that test

Reply:
What does that mean? I don't see anything in the old code about this

Reply:
We automatically detect where servers are running based on runtime configuration. Typically, 1 GE server per host where workers run with GE present on a worker with `local_rank = 0`. This unit test lists possible scenarios:
https://github.com/microsoft/DeepGNN/blob/main/src/python/deepgnn/graph_engine/backends/snark/test_snark_client.py

Reply:
So if a user sets "OMPI_COMM_WORLD_LOCAL_RANK": "0",
as an environment variable we automatically set the hostname to the horovod default IP?

Reply:
yes